test/test_emulation/__init__.py

```diff
@@ -0,0 +1,168 @@
+
+    def send_event(self, id, event):
+        response = self._post(f'/devices/{id}/event', json=event)
+        assert response.status_code == 200
+
+
+class TestDevice(object):
+    def reload_udev_rules(self):
+        subprocess.run('udevadm control --reload-rules'.split())
+
+    @pytest.fixture(scope='session', autouse=True)
+    def udev_rules(self):
+        rules_dir = Path('/run/udev/rules.d')
+        rules_dir.mkdir(exist_ok=True)
+        shutil.copyfile('test/test_emulation/rules.d/61-ratbag-emu-ignore-test-devices.rules', '/run/udev/rules.d/61-ratbag-emu-ignore-test-devices.rules')
+        self.reload_udev_rules()
+
+        yield
+
+        rules = rules_dir.joinpath('61-ratbag-emu-ignore-test-devices.rules')
+        if rules.is_file():
+            rules.unlink()
+            self.reload_udev_rules()
+
+    @pytest.fixture(autouse=True, scope='session')
+    def server(self):
+        p = None
+        try:
+            p = subprocess.Popen('python -m ratbag_emu 2&>1 >/dev/null', shell=True)
+            sleep(2)
+            yield
+        finally:
+            if p:
+                p.kill()
+
+    @pytest.fixture(autouse=True, scope='session')
+    @pytest.mark.usesfixtures('server')
+    def client(self):
+        yield RatbagemuClient()
+
+    @pytest.fixture
+    def id(self, client):
+        id = client.create(self.shortname)
+
+        yield id
+
+        client.delete(id)
+
+    def _simulate(self, client, id, data):
+        input_nodes = client.get_input_nodes(id)
+
+        # Open the event nodes
+        event_nodes = []
+        for node in set(input_nodes):
+            fd = open(node, 'rb')
+            fcntl.fcntl(fd, fcntl.F_SETFL, os.O_NONBLOCK)
+            event_nodes.append(libevdev.Device(fd))
+
+        events = []
+        def collect_events(stop):
+            nonlocal events
+            while not stop.is_set():
+                for node in event_nodes:
+                    events += list(node.events())
+
+        stop_event_thread = threading.Event()
+        event_thread = threading.Thread(target=collect_events, args=(stop_event_thread,))
+        event_thread.start()
+
+        client.send_event(id, data)
+
+        sleep(1)
+        stop_event_thread.set()
+        event_thread.join()
+
+        for node in event_nodes:
+            node.fd.close()
+
+        received = MouseData()
+        for e in events:
+            if e.matches(libevdev.EV_REL.REL_X):
+                received.x += e.value
+            elif e.matches(libevdev.EV_REL.REL_Y):
+                received.y += e.value
+
+        return received
+
+
+class MouseData(object):
+    '''
+    Holds event data
+    '''
+
+    def __init__(self, x=0, y=0):
+        self.x = x
+        self.y = y
+
+    @staticmethod
+    def from_mm(dpi, x=0, y=0):
+        return MouseData(x=int(x * MM_TO_INCH * dpi),
+                         y=int(y * MM_TO_INCH * dpi))
+
+
+MM_TO_INCH = 0.0393700787
```

Comment:
maybe I'm just old-fashioned here but constants usually go at the top of the file :)